bootstore/src/schemes/v0/fsm.rs

```diff
@@ -0,0 +1,820 @@
+            ResponseType::Pkg(pkg) => self.on_pkg(from, rsp.request_id, pkg),
+            ResponseType::Error(error) => {
+                Err(ApiError::ErrorResponseReceived {
+                    from,
+                    state: self.state.name(),
+                    request_id: rsp.request_id,
+                    error,
+                })
+            }
+        }
+    }
+
+    // Handle a `ResposneType::InitAck` from a peer
+    fn on_init_ack(
+        &mut self,
+        from: Baseboard,
+        request_id: Uuid,
+    ) -> Result<Option<ApiOutput>, ApiError> {
+        match self.request_manager.on_init_ack(from.clone(), request_id) {
+            Some(true) => Ok(Some(ApiOutput::RackInitComplete)),
+            Some(false) => Ok(None),
+            None => Err(ApiError::UnexpectedResponse {
+                from,
+                state: self.state.name(),
+                request_id,
+                msg: "InitAck",
+            }),
+        }
+    }
+
+    // Handle a `ResponseType::Pkg` from a peer
+    fn on_pkg(
+        &mut self,
+        from: Baseboard,
+        request_id: Uuid,
+        pkg: LearnedSharePkg,
+    ) -> Result<Option<ApiOutput>, ApiError> {
+        if self.request_manager.on_pkg(request_id) {
+            // This pkg matched our outstanding request. Let's transition from
+            // `State::Learning` to `State::Learned`.
+            assert_eq!(self.state, State::Learning);
+            self.state = State::Learned { pkg };
+            Ok(Some(ApiOutput::LearningCompleted))
+        } else {
+            if self.state == State::Learning {
+                // This is a stale response. We could choose to accept it, but
+                // for consistency with the rest of the `TrackableRequests`
+                // we  only accept responses that have currently outstanding
+                // requests.
+                Ok(None)
+            } else {
+                Err(ApiError::UnexpectedResponse {
+                    from,
+                    state: self.state.name(),
+                    request_id,
+                    msg: "Pkg",
+                })
+            }
+        }
+    }
+
+    // Handle a `ResponseType::Share` from a peer
+    fn on_share(
+        &mut self,
+        from: Baseboard,
+        request_id: Uuid,
+        share: Share,
+    ) -> Result<Option<ApiOutput>, ApiError> {
+        match &mut self.state {
+            // We don't send `GetShare` requests in these states
+            State::Uninitialized | State::Learning => {
+                return Err(ApiError::UnexpectedResponse {
+                    from,
+                    state: self.state.name(),
+                    request_id,
+                    msg: "Share",
+                });
+            }
+            State::InitialMember { pkg, distributed_shares } => {
+                validate_share(&from, &share, &pkg.share_digests)?;
+                match self.request_manager.on_share(from, request_id, share) {
```

Comment:
API-wise, this seems weird to me – it feels like leakage of `RequestManager` details.  I'd expect `on_share` to be opaque (returning a `Result<Option<ApiOutput>, ApiError> `), since the request manager knows what it wants to do next.

This would require passing `&mut self.envelopes` into `request_manager.on_share(..)`, which seems fine to me.

Reply:
While I get your point, I respectfully disagree with the proposed solution. While the `RequestManager` is exposing some internal types here, it's knowledge of `Fsm` types and logic in this file is extremely limited. `RequestManager` currently doesn't know about `ApiOutput`, `ApiError` or about decryption algorithms, combining shares, etc.. The sole purview of `RequestManager`, is tracking outstanding requests and ensuring that all of them appropriately expire or complete given their constraints, and that necessary request retries are sent when peers are connected.

Reply:
Okay, I'm fine leaving the `RequestManager` stuff as-is; thanks for explaining the architecture.